src/gafaelfawr/services/userinfo.py

```diff
@@ -0,0 +1,470 @@
+"""Service and caching layer for user metadata."""
+
+from __future__ import annotations
+
+import re
+from typing import List, Optional
+
+from structlog.stdlib import BoundLogger
+
+from ..config import OIDCConfig
+from ..constants import BOT_USERNAME_REGEX
+from ..dependencies.cache import IdCache
+from ..exceptions import (
+    InvalidTokenClaimsException,
+    MissingClaimsException,
+    ValidationError,
+)
+from ..models.oidc import OIDCVerifiedToken
+from ..models.token import TokenData, TokenGroup, TokenUserInfo
+from ..storage.firestore import FirestoreStorage
+from ..storage.ldap import LDAPStorage
+
+__all__ = ["OIDCUserInfoService", "UserInfoService"]
+
+
+class UserInfoService:
+    """Retrieve user metadata from external systems.
+
+    In some cases, we take user metadata from external systems.  Examples are:
+
+    #. Resolve a unique identifier to a username via LDAP.
+    #. Get user group membership from LDAP.
+    #. Get UID or GID from LDAP.
+    #. Assign and manage UIDs and GIDs via Google Firestore.
+
+    This service manages those interactions.  UID/GID data from Firestore is
+    cached.  LDAP data is not cached, since LDAP is supposed to be able to
+    handle a very high query load.
```

Comment:
Ah, so the caching is not about latency to the data but about respecting the backend service?

Reply:
This comment is probably inaccurate on further reflection, and I'm going to put in a five-minute cache for all LDAP information. I'm not worried about the latency during initial login, since that's going to be kind of slow anyway, but if the user is making a flurry of authenticated API requests, we probably need to prevent that from turning into a flurry of LDAP queries, particularly since we don't maintain a persistent connection right now. I'll do that in a separate PR.